File: xuri-rpc-java/src/main/java/com/xuri/rpc/ObjectOfProxyManager.java
package com.xuri.rpc;

import java.util.HashMap;
import java.util.IdentityHashMap;
import java.util.Iterator;
import java.util.Map;

/**
 * 本地对象代理管理器，管理本地对象与代理ID的双向映射
 * 对应TS中: export class ObjectOfProxyManager
 *
 * proxyMap使用IdentityHashMap，等价于TS中以对象引用为key的Map
 */
public class ObjectOfProxyManager {

    /**
     * 代理对象处理器，记录代理ID、目标对象和最后注册时间
     * 对应TS中: class ProxyObjectHandlerForManager
     */
    static class ProxyObjectHandler {
        final String id;
        final Object target;
        long lastRegistered;

        ProxyObjectHandler(String id, Object target) {
            this.id = id;
            this.target = target;
            this.lastRegistered = System.currentTimeMillis();
        }
    }

    // 对象 → ID (使用引用相等, 等价于TS中的Map<object, string>)
    private final Map<Object, String> proxyMap = new IdentityHashMap<>();

    // ID → Handler
    private final Map<String, ProxyObjectHandler> reverseProxyMap = new HashMap<>();

    public void set(Object obj, String id) {
        proxyMap.put(obj, id);
        reverseProxyMap.put(id, new ProxyObjectHandler(id, obj));
    }

    public void reRegister(String id) {
        ProxyObjectHandler handler = reverseProxyMap.get(id);
        if (handler != null) {
            handler.lastRegistered = System.currentTimeMillis();
        }
    }

    public Object getById(String id) {
        ProxyObjectHandler handler = reverseProxyMap.get(id);
        return handler != null ? handler.target : null;
    }

    public String get(Object obj) {
        return proxyMap.get(obj);
    }

    public boolean has(Object obj) {
        return proxyMap.containsKey(obj);
    }

    public void deleteById(String id) {
        ProxyObjectHandler handler = reverseProxyMap.get(id);
        if (handler != null) {
            proxyMap.remove(handler.target);
        }
        reverseProxyMap.remove(id);
    }

    public void delete(Object obj) {
        String id = proxyMap.get(obj);
        if (id != null) {
            reverseProxyMap.remove(id);
        }
        proxyMap.remove(obj);
    }

    public int size() {
        return proxyMap.size();
    }

    /**
     * 移除超时的代理对象
     * 对应TS中 removeOutdatedProxyObject 函数对 reverseProxyMap 的遍历逻辑
     *
     * @param timeoutMs 超时阈值(毫秒)，实际使用 timeout*3
     * @return 被移除的数量
     */
    public int removeOutdated(long timeoutMs) {
        int count = 0;
        Iterator<Map.Entry<String, ProxyObjectHandler>> it = reverseProxyMap.entrySet().iterator();
        while (it.hasNext()) {
            Map.Entry<String, ProxyObjectHandler> entry = it.next();
            if (System.currentTimeMillis() - entry.getValue().lastRegistered > timeoutMs * 3) {
                proxyMap.remove(entry.getValue().target);
                it.remove();
                count++;
            }
        }
        return count;
    }
}

File: xuri-rpc-java/src/main/kotlin/com/xuri/rpc/ObjectOfProxyManager.kt
package com.xuri.rpc

import java.util.IdentityHashMap

/**
 * 本地对象代理管理器，管理本地对象与代理ID的双向映射
 * 对应TS中: export class ObjectOfProxyManager
 *
 * proxyMap 使用 IdentityHashMap，等价于TS中以对象引用为key的Map
 */
class ObjectOfProxyManager {

    /**
     * 代理对象处理器
     * 对应TS中: class ProxyObjectHandlerForManager
     */
    private class ProxyObjectHandler(
        val id: String,
        val target: Any,
        var lastRegistered: Long = System.currentTimeMillis()
    )

    // 对象 → ID (引用相等)
    private val proxyMap = IdentityHashMap<Any, String>()

    // ID → Handler
    private val reverseProxyMap = mutableMapOf<String, ProxyObjectHandler>()

    fun set(obj: Any, id: String) {
        proxyMap[obj] = id
        reverseProxyMap[id] = ProxyObjectHandler(id, obj)
    }

    fun reRegister(id: String) {
        reverseProxyMap[id]?.let { it.lastRegistered = System.currentTimeMillis() }
    }

    fun getById(id: String): Any? = reverseProxyMap[id]?.target

    fun get(obj: Any): String? = proxyMap[obj]

    fun has(obj: Any): Boolean = obj in proxyMap

    fun deleteById(id: String) {
        reverseProxyMap.remove(id)?.let { proxyMap.remove(it.target) }
    }

    fun delete(obj: Any) {
        proxyMap.remove(obj)?.let { reverseProxyMap.remove(it) }
    }

    val size: Int get() = proxyMap.size

    /**
     * 移除超时的代理对象
     * 对应TS中 removeOutdatedProxyObject 对 reverseProxyMap 的遍历 (已修复bug)
     *
     * @param timeoutMs 超时阈值(毫秒)，实际比较用 timeout*3
     * @return 被移除的数量
     */
    fun removeOutdated(timeoutMs: Long): Int {
        var count = 0
        val now = System.currentTimeMillis()
        val iter = reverseProxyMap.entries.iterator()
        while (iter.hasNext()) {
            val entry = iter.next()
            if (now - entry.value.lastRegistered > timeoutMs * 3) {
                proxyMap.remove(entry.value.target)
                iter.remove()
                count++
            }
        }
        return count
    }
}
